### DISCUSSION TITLE
Porting to STM32CubeIDE

### DISCUSSION THREAD

**Discussion by kvkc97** [Lora-net/LoRaMac-node#1160] [Q&A]
I would like to port LoRaWAN Stack for a custom board built using STM32WB55RGV6 controller and LR1110 driver. I am using STM32CubeIDE for the same. Can I port and use the stack in STM32CubeIDE?

**Comment by Pirakutu** [comment]
yes you can, i have a version of the stack running on a stm32wb55, with the LR1110, on CubeIDE. It works well.

**Reply by wrig0421** [reply]
Which version of LoRaMAC node are you using?  I am also using a STM32WB55.  I was able to successfully port and develop on basicmac version of stack (built off IBMs LMIC).  I'm trying to get v4.4.7 working but I'm not having much luck getting consistent uplinks on gateway.  I had to make a hack in the code also it seems after the first uplink no additional uplinks occur unless I force the MACDone flag to true following TXDONE interrupt.  Not sure if you ran into any similar issues.

**Reply by kvkc97** [reply]
I am following the latest version of LoRaMAC node. I think I am having the same issue as yurs (MACDone flag not being set). Further uplinks are not happening. How to fix it

**Comment by kvkc97** [comment]
@Pirakutu I have ported the stack to CubeIDE to work with STM32WB55RG controller. The RF communication was checked with LR1110 IC and the controller and packet was received at the gateway. However, after including the MAC layer in the code, the code gets stuck at SPI transmission  command. Could you please share your code so that I can use it as a reference?

**Comment by Pirakutu** [comment]
What do you mean after including the mac layer? 

Does it get stuck at the busy pin or at an SPI command? As these are 2 different things.

**Comment by kvkc97** [comment]
I have tested the LR1110 IC with STM32WB55RG controller using STM32CubeIDE and packet was received at the gateway. Afterwords I added Mac layer ( Mac, peripherals, LoRaWAN folder from apps) and made necessary changes in boards folder for STM32WB55RG controller and called the Lmhandler functions in main.cpp as given in this example. On running the code, it gets stuck at SPI transmission (it works well when communicating directly using IC). Busy pin is working fine and is set low as required.

**Comment by kvkc97** [comment]
I have updated the code for SPI and the code is not getting stuck. However some other errors are displayed.
I was testing for LoRa transmission with LR1110 IC in OTA mode using LoRaMAC code. The Join request is successfully sent to the server and Join accept is generated from the server and is passed to the gateway. The gateway also forwards the packet. However the packet is not received at the end device (custom board developed using STM32 controller). I have attached the Join request and join accept messages.
The IC is going to rx mode after transmitting but packet is not received. At time the IC goes to busy after TX mode and gets stuck. At times it goes to RX mode. A few times when in RX mode, interrupt for preamble detection is generated but no further packet is received and most of the times preamble detection also is not done. What would be the issue? Is it due to the PLL error generated.
Join request at server
![image](https://user-images.githubusercontent.com/73599382/132336158-30856182-d7a5-481f-9df5-eb4035deaa66.png)
Join accept at server
![image](https://user-images.githubusercontent.com/73599382/132336220-f1277f85-52a4-4651-a52d-7e698d805b68.png)
Join request and Join Request at gateway
![image](https://user-images.githubusercontent.com/73599382/132336275-121dc652-5a1d-422c-8454-3e9cb56746df.png)

**Comment by Pirakutu** [comment]
which LR1110 driver files did you use? i had problems using the newest ones. Had to go back to version 1.0.0.

**Comment by kvkc97** [comment]
I used the latest version given in the following link
https://github.com/Lora-net/lr1110_driver
Did you go back to version 1.0.0 of the same driver?

**Comment by Pirakutu** [comment]
Yeah, i went back to the version 1.0.0 of that same driver (i saw that was the version included in this project and hoped it worked that way).

**Comment by kvkc97** [comment]
Hi, I reverted to version 1.0.0 and tried the code. The errors are still present. 
Also I have opened an issue in LR1110 Driver github page regarding the PLL error. The link is https://github.com/Lora-net/lr1110_driver/issues/16

**Comment by kvkc97** [comment]
The firmware version in LR1110 is 0303. Is this version compatible with LR1110 Version1.0.0. If not, how can I change the firmware version of LR1110? Is it possible using SPI?
